lib/Sema/TypeChecker.h

```diff
@@ -465,10 +465,49 @@ RequirementCheckResult checkGenericArguments(
     ArrayRef<Requirement> requirements,
     TypeSubstitutionFn substitutions);
 
-bool checkContextualRequirements(GenericTypeDecl *decl,
-                                 Type parentTy,
-                                 SourceLoc loc,
-                                 DeclContext *dc);
+/// Checks whether the generic requirements imposed on the nested type
+/// declaration \p decl (if present) are in agreement with the substitutions
+/// that are necessary to spell it within the given parent type \p parentTy,
+/// when mapped into context using the given generic environment.
+///
+/// For example, given
+/// \code
+/// struct S<X> {}
+/// extension S where X == Bool {
+///   struct Inner {}
+/// }
+/// \endcode
+/// \c Inner cannot be referenced on \c S<Int>, because its contextual
+/// requirement \c X \c == \c Bool is not satisfied by the substitution
+/// \c [X \c = \c Int].
+///
+/// Similarly, \c typealias \c Y below is a viable type witness in the
+/// conformance of \c S to \c P, because its contextual requirement
+/// \c Self.X \c == \c Bool is satisfied by the substitution
+/// \c [Self \c = \c S].
+/// \code
+/// protocol P {
+///   associatedtype X
+///   associatedtype Y
+/// }
+/// extension P where X == Bool {
+///   typealias Y = Bool
+/// }
+///
+/// struct S: P {
+///   typealias X = Bool
+/// }
+/// \endcode
+///
+/// \param module The module to use for conformace lookup.
+/// \param getGenericEnvironment A function object that enables on-demand
+/// computation of the generic environment used to map the parent type into
+/// context.
+///
+/// \returns \c true on success.
+bool checkContextualRequirements(
+    GenericTypeDecl *decl, Type parentTy, SourceLoc loc, ModuleDecl *module,
+    llvm::function_ref<const GenericEnvironment *()> getGenericEnvironment);
```

Comment:
Instead, can you pass a GenericSignature and call getGenericEnvironment() on it when needed?

Reply:
Ah, great suggestion!

Reply:
Passing a GenericSignature instead of a function ref results in one more duplicate circularity error in [compiler_crashers_2_fixed/0161-sr6569.swift](https://github.com/apple/swift/pull/39652/files#diff-be9258549c9eed8f157ce4ccd079e2e5a3fd5c0ab703e319380e7cfc8b1e9b49). Do you reckon that sounds fine? All this fuss around checkContextualRequirements() is really only to make sure that is works correctly with SIL code, where we explicitly pass a generic environment to type resolution, because it cannot be derived from context.